### PR TITLE
[experiment] feat: migrate to fastify

### PR DESCRIPTION
**Pull request by clarkdo**
Please use #4939 issue for discussion.

This is a proposal pull request for migrating connect to fastify, it's still WIP.

- [ ] If it's necessary and feasible to support `connect` and `fastify` at same time by dynamic import different server module like `server-connect` or `server-fastify`
- [ ] The future of custom server and programmatic usage
- [ ] serverMiddleware supports [fastify route](https://www.fastify.io/docs/latest/Routes/)

### CODE REVIEW

File: packages/server/src/listener.js [hunk 1]
```diff
@@ -17,29 +13,25 @@ export default class Listener {
...
     this.address = null
     this.url = null
   }
 
   async close() {
     // Destroy server by forcing every connection to be closed
-    if (this.server && this.server.listening) {
-      await this.server.destroy()
+    if (this.app.server && this.app.server.listening) {
+      await this.close()
```
Comment (manniL) [inline packages/server/src/listener.js]: wait for `this.close` inside `this.close`? :thinking:

**Comment by galvez** [comment]
It would be nice if we could have built-in JSON reading / writing. I'm currently using this:

```
import express from 'express'

const jsonMiddleware = express.json()

export default function(req, res, next) {
  res.json = (obj) => {
    res.write(JSON.stringify(obj))
  }
  return jsonMiddleware(req, res, next)
}
```

added to my `serverMiddleware`.

**Comment by sh** [comment]
>  If it's necessary and feasible to support connect and fastify at same time by dynamic import different server module like server-connect or server-fastify

We're running different services in our stack that all can share common middlewares (session, csrf) thanks to nuxt's connect compatibility.

We're also using nuxt as middleware in our express server. This allows us to have automatic request tracing through [GCP Stackdriver Trace Agent for Node.js](https://github.com/googleapis/cloud-trace-nodejs#stackdriver-trace-agent-for-nodejs) which is a massive help for debugging and measuring issues in production. As it stands there is no fastify tracing plugin available, migrating to fastify (without the option of still using connect) would be a massive drawback for us.

I'm not sure how much of a gain fastify (besides having a bit more req/s in benchmarks compared to connect/connect-router) really would be as it breaks the compatibility with the proven connect/express ecosystem.

I'd vote for support of both frameworks (connect+fastify) or stay with connect if this change introduces too much complexity.

**Comment by Atinux** [comment]
After talking with the team about it, I think we can close this one directly.

Thank you for your great feedback @sh

**Pull request closed without merging**